src/Dialect/ONNX/DisposablePool.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------------------------- DisposablePool.hpp ------------------------===//
+//
+// DisposablePool manages instances of DisposableElementsAttr.
+// It creates them, maintains a record of them (in a "pool") until they are
+// deemed unreachable, and it can be called to garbage collect those that are
+// unreachable, and to "scrub" all occurrences in a module by replacing each
+// with a DenseElementsAttr.
+//
+// Garbage collected and scrubbed DisposableElementsAttrs are removed from the
+// pool and their reference to the underlying MemoryBuffer is cleared,
+// decrementing the shared_ptr reference count.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "src/Dialect/ONNX/DisposableElementsAttr.hpp"
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/DialectInterface.h"
+
+#include <unordered_map>
+
+namespace onnx_mlir {
+
+class DisposablePool : public mlir::DialectInterface::Base<DisposablePool> {
+  friend class ElementsAttrBuilder; // allow access to insert()
+public:
+  static DisposablePool &create(mlir::MLIRContext *context);
+
+  static DisposablePool *get(mlir::MLIRContext *context);
+
+  // Disposes every DisposableElementsAttr and in moduleOp replaces each with a
+  // DenseElementsAttr.
```

Comment:
Can you comment in the comment if that process is reversible, or meant to be used to permanently go toward regular dense element attributes. Tx

Reply:
I added the sentence:
```
This is irreversible and is called when we are done transforming the ONNX dialect, just before we lower it.
```